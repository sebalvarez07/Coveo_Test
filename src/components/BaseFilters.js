import React from 'react';
import { connect } from 'react-redux';
import { startUpdateBaseFilter } from '../actions/filters';

export class BaseFilters extends React.Component {

    constructor (props) {
        super(props);
        this.beerButton = React.createRef();
        this.wineButton = React.createRef();
        this.state = {
            beerActive: true,
            wineActive: true
        }
    }

    componentDidUpdate (prevProps, prevState) {
        if(prevState.beerActive !== this.state.beerActive) {
            this.beerButton.current.classList.toggle('active-base');
        }
        
        if(prevState.wineActive !== this.state.wineActive) {
            this.wineButton.current.classList.toggle('active-base');
        }
    }
    
    dispatchBaseFilter = () => {
        let baseFilter;
        const beerActive = this.state.beerActive;
        const wineActive = this.state.wineActive;

        if(beerActive === wineActive) {
            baseFilter = ['Bière', 'Vin'];
        }
        else if(beerActive && !wineActive) {
            baseFilter = ['Bière'];
        }
        else if(!beerActive && wineActive) {
            baseFilter = ['Vin'];
        }

        this.props.startUpdateBaseFilter(baseFilter);
    }

    handleBaseFilter = (e) => {
        e.preventDefault(); 
        const btn = e.target.closest('.base-filter__btn');

        if(btn.id === 'beerFilter') {
            this.setState(prevState => (
                {
                    beerActive: !prevState.beerActive
                }
                ), this.dispatchBaseFilter);
        }
        else {
            this.setState(prevState => (
                {
                    wineActive: !prevState.wineActive
                }
                ), this.dispatchBaseFilter);
        }
    }

    render () {
        return (
            <div className='base-filter header__el'>
                <span 
                    ref={this.beerButton}
                    className='base-filter__btn active-base'
                    id={'beerFilter'}
                    onMouseDown={this.handleBaseFilter}
                >
                    <i className="ionicons ion-ios-pint"></i>
                    <span className='base-filter__label'>
                        Beer
                    </span>
                </span>
                <span
                    ref={this.wineButton}
                    className='base-filter__btn active-base'
                    id={'wineFilter'}
                    onMouseDown={this.handleBaseFilter}
                >
                    <i className="ionicons ion-ios-wineglass"></i>
                    <span className='base-filter__label'>
                        Wine
                    </span>
                </span>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startUpdateBaseFilter: (baseFilter) => dispatch(startUpdateBaseFilter(baseFilter))
    }
}
 
export default connect(undefined, mapDispatchToProps)(BaseFilters);