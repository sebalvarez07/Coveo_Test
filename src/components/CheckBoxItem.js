import React, { useState } from 'react';

const CheckBoxItem = (props) => {

};

export default CheckBoxItem;