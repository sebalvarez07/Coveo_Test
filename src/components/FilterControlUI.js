import React from 'react';
import { connect } from 'react-redux';
import { groupMap } from '../fixtures/helpers';
import { startRemoveFilterValue } from '../actions/filters';

export const FilterControlUI = (props) => {

    const handleValueRemoval = (e) => {
        const btn = e.target.closest('.filter-control__btn');
        const group = btn.dataset.group;
        const val = btn.dataset.value;
        props.startRemoveFilterValue(group, val);
    } 

    return (
        <div className='filter-controlUI'>
            <h4 className='current-viewing'>
                Currently Viewing: {props.baseFilter.join(' & ')}
            </h4>
            {
                Object.keys(props.filtersActive).map(category => {
                    if(props.filtersActive[`${category}`].values.length > 0) {
                        return (
                            <div className='filter-control__container' key={category}>
                                <span className='filter-control__cat'>{groupMap.get(category)}: </span>
                                {
                                    props.filtersActive[category].values.map(activeFilterVal => {
                                        return <button 
                                                    data-group={category}
                                                    data-value={activeFilterVal}
                                                    className='filter-control__btn'
                                                    onMouseDown={handleValueRemoval}
                                                    key={activeFilterVal}>
                                                        {activeFilterVal}
                                                        <i className="ionicons ion-close"></i>
                                                </button>
                                    })
                                }
                            </div>
                        )    
                    }
                })
            }
        </div>
    )
}

const mapStateToProps = (state) => (
    {
        filtersActive: state.filters.categories,
        baseFilter: state.filters.baseFilter
    }
);

const mapDispatchToProps = (dispatch) => (
    {
        startRemoveFilterValue: (group, value) => dispatch(startRemoveFilterValue(group, value)),
    }
);

export default connect(mapStateToProps, mapDispatchToProps)(FilterControlUI);