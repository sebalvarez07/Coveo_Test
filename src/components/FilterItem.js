import React from 'react';
import { connect } from 'react-redux';
import { startUpdateCategory, startUpdateCategoryLimit } from'../actions/filters';
import { groupMap } from '../fixtures/helpers';

const defaultLimit = 5;

class FilterItem extends React.Component {

    constructor(props) {
        super(props);
        this.checkBoxes = [];
        this.checkedItems = new Set();
        this.state = {
            valueLimit: defaultLimit
        }
    }

    componentWillMount () {
        this.checkBoxes = this.props.filter.map(option => option.value);
    }

    componentWillUpdate (nextProps) {

        this.checkedItems = new Set([...this.props.checkedFilters]);

        if(nextProps.baseFilter !== this.props.baseFilter) {
            // TODO.. on baseFilter change, remove wine || beer filters           
        }
        
        // Ensure to keep check boxes on view
        this.checkBoxes = [...this.checkedItems];
    
        // Keep old checked boxes, and add new boxes
        nextProps.filter.forEach(option => {
            if(!this.checkedItems.has(option.value)){
                this.checkBoxes.push(option.value)
            }
        });     
    }

    handleChange = (e) => {
        const item = e.target.name;
        const isChecked = e.target.checked;
            
        if(isChecked) {
            this.checkedItems.add(item)
        } else {
            if(this.checkedItems.has(item)) {
                this.checkedItems.delete(item);
            }
        }

        const checkedItemsArr = Array.from(this.checkedItems);

        this.props.startUpdateCategory(this.props.filterName, checkedItemsArr);
    }

    dispatchMoreValues = () => {
        if(this.state.valueLimit <= this.props.filter.length) {
            return;
        } else {
            this.props.startUpdateCategoryLimit(this.props.filterName, this.state.valueLimit);
        }
    }

    handleValueLimit = (e) => {
        e.preventDefault();
        
        if(e.target.name === 'more'){
            this.setState(prevState => {
                return {
                    valueLimit: prevState.valueLimit + 5
                }
            }, this.dispatchMoreValues)
        }

        else if(e.target.name === 'less'){
            const resetLimit = this.checkedItems.size > defaultLimit ? this.checkedItems.size : defaultLimit;
            this.setState({ valueLimit: resetLimit});
        }
    }

    handleClearAll = () => {
        this.props.startUpdateCategory(this.props.filterName, []);
    }

    setLimit = () => {
        
        return this.checkedItems.size > this.state.valueLimit ? this.checkedItems.size : this.state.valueLimit;
    }

    renderCheckBox = (checkName) => {        
        return (
            <label className='check__container' key={`${checkName}`}>
                {`${checkName}`}
                <input 
                    className='custom__checkbox'
                    style={{ display: 'block' }} 
                    type='checkbox' 
                    name={`${checkName}`} 
                    checked={this.checkedItems.has(checkName)}
                    onChange={this.handleChange} 
                />
                <span className='checkmark'>
                    <i className="ionicons ion-checkmark"></i>
                </span>
            </label>
        )
    }
    
    render (){
        const limit = this.setLimit();
        return (
            <div className='sidebar__group'>
                <div className='group__header'>
                    <h3 className='group__tl'>{groupMap.get(this.props.filterName)}</h3>
                    <span className='clear-btn' onMouseDown={this.handleClearAll}>Clear All</span>
                </div>
                
                {
                    this.checkBoxes.reduce((acc, checkName, index) => {
                        if(limit > index) {
                            acc.push(this.renderCheckBox(checkName))
                        }
                        return acc;
                    }, [])
                }
                
                <div className='more_less__group'>
                    { this.state.valueLimit <= this.props.filter.length && 
                        <React.Fragment>
                            <button 
                                className='more_less__btn'
                                name={'more'}
                                onMouseDown={this.handleValueLimit}
                                >
                                Show More
                            </button>
                            <span> | </span>
                        </React.Fragment>
                    }
                    { this.props.filter.length >= defaultLimit && 
                        <button 
                            className='more_less__btn'
                            name={'less'}
                            onMouseDown={this.handleValueLimit}
                            >
                            Show Less
                        </button>
                    }
                </div>
                
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startUpdateCategory: (group, update) => dispatch(startUpdateCategory(group, update)),
        startUpdateCategoryLimit: (group, valueLimit) => dispatch(startUpdateCategoryLimit(group, valueLimit))
    }
}

const mapStateToProps = (state, props) => {
    return  {
        baseFilter: state.filters.baseFilter,
        checkedFilters: state.filters.categories[`${props.filterName}`]['values']
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(FilterItem);