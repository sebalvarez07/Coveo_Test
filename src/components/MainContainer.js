import React from 'react';
import { connect } from 'react-redux';
import ResultCard from './ResultCard';
import Pagination from './Pagination';
 
const MainContainer = (props) => {

    const showingResults = `Showing results ${props.firstResult} - ${props.firstResult + props.numberOfResults}`;

    return (
        <div className='main-container'>
            <div className='results-info'>
                <div>Total number of results <strong>{props.totalResults}</strong></div>
                <div>{showingResults}</div>
                <Pagination />
            </div>
            <div className='card-container'>
                {
                    props.results.map(result => {
                        return (
                            <ResultCard key={result.uniqueId} result={result}/>
                        )
                    })
                }
            </div>
            <Pagination />
        </div>
    )
}

const mapStateToProps = (state) => (
    {
        totalResults: state.content.resultsCount,
        numberOfResults: state.filters.numberOfResults,
        firstResult: state.filters.firstResult,
        results: state.content.results,
    }
);

export default connect(mapStateToProps)(MainContainer);