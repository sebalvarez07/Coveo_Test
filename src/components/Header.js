import React from 'react';
import BaseFilters from './BaseFilters';
import SearchBar from './SearchBar';
import PriceSlider from './PriceSlider';

export const Header = () => {
    return (
        <header className="header">
            <div className="header-container">
                <div className="logo-container">
                    <span className='logo'>coveo</span>
                </div>
                <div className="header__content">
                    <BaseFilters />
                    <PriceSlider />
                    <SearchBar />
                </div>
            </div>
        </header>
    );   
}

export default Header;