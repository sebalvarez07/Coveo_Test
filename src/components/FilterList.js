import React from 'react';
import { connect } from 'react-redux';
import FilterItem from './FilterItem';
 
export class FilterList extends React.Component {
    constructor(props) {
        super(props)
    }

    render () {
        return (
            <div>
                {
                    this.props.filters.map(filter => {
                        return (
                            <FilterItem key={`${filter.field}`} filterName={filter.field} filter={filter.values}/>
                        )
                    })
                }
            </div>
        )
    }
 }

 const mapStateToProps = (state) => (
    {
        filters: state.content.groupBy
    }
);

 export default connect(mapStateToProps)(FilterList);