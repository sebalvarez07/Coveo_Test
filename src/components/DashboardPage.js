import React from 'react';
import Header from './Header';
import Sidebar from './Sidebar';
import MainContainer from './MainContainer';
 
const DashboardPage = (props) => {
    return (
        <div className='wrapper'>
            <Header />
            <Sidebar />
            <MainContainer />
        </div>
    )
}

export default DashboardPage;