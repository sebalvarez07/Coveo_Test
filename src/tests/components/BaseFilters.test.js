import React from 'react';
import { shallow } from 'enzyme';
import { BaseFilters } from '../../components/BaseFilters';

let startUpdateBaseFilter, wrapper;

beforeEach( () => {
    startUpdateBaseFilter = jest.fn();
    wrapper = shallow(<BaseFilters startUpdateBaseFilter={startUpdateBaseFilter} />)
});

test('Should render base filters', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle startUpdateBaseFilter when beer filter is clicked', () => {
    // simulate mouseDown and pass the fake element's object 
    wrapper.find('#beerFilter').simulate('mouseDown', {
        preventDefault: () => {},
        target: {
            closest: (btn) => {
                return {id: 'beerFilter'}
            }
        }
    });

    expect(startUpdateBaseFilter).toHaveBeenLastCalledWith(['Vin']);
});

test('Should handle startUpdateBaseFilter when wine filter is clicked', () => {
    // simulate mouseDown and pass the fake element's object 
    wrapper.find('#wineFilter').simulate('mouseDown', {
        preventDefault: () => {},
        target: {
            closest: (btn) => {
                return {id: 'wineFilter'}
            }
        }
    });

    expect(startUpdateBaseFilter).toHaveBeenLastCalledWith(['Bi√®re']);
});

