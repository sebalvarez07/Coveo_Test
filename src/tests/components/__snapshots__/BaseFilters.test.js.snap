// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render base filters 1`] = `
<div
  className="base-filter header__el"
>
  <span
    className="base-filter__btn active-base"
    id="beerFilter"
    onMouseDown={[Function]}
  >
    <i
      className="ionicons ion-ios-pint"
    />
    <span
      className="base-filter__label"
    >
      Beer
    </span>
  </span>
  <span
    className="base-filter__btn active-base"
    id="wineFilter"
    onMouseDown={[Function]}
  >
    <i
      className="ionicons ion-ios-wineglass"
    />
    <span
      className="base-filter__label"
    >
      Wine
    </span>
  </span>
</div>
`;
